In [25]:
from pathlib import Path
import sys

# Get the project root directory
PROJECT_ROOT = Path.cwd().parent

# Add it to Python path
sys.path.append(str(PROJECT_ROOT))

In [26]:
import numpy as np 
import pandas as pandas

In [27]:
from src.data.loader import load_csv
from src.data.constants import RAW_DATA_DIR

In [28]:
movies = load_csv(RAW_DATA_DIR / "movielens" / "movies.csv")
ratings = load_csv(RAW_DATA_DIR / "movielens" / "ratings.csv")
links = load_csv(RAW_DATA_DIR / "movielens" / "links.csv")
tags = load_csv(RAW_DATA_DIR / "movielens" / "tags.csv")

In [ ]:
print("=="*50)
print("Movies Dataset")
print("=="*50)

print("First 5 Rows : ")
display(movies.head())
print("\nBasic Info : \n",movies.info())
print("Shape : ",movies.shape)
print("\nColumns : ",movies.columns.tolist())
print("\nDataTypes : \n",movies.dtypes)
print("\nMissing Value : \n",movies.isnull().sum())
print("\nDuplicate Rows : ",movies.duplicated().sum())
print("\nNo. of Genres : ",movies['genres'].nunique())
print("\nGenres type : ",movies["genres"].str.split("|").explode().unique().tolist())
movies["genres"].str.split("|").explode().value_counts()


Movies Dataset
First 5 Rows : 


,movieId,title,genres
0,1,Toy Story (1995),Adventure|Animation|Children|Comedy|Fantasy
1,2,Jumanji (1995),Adventure|Children|Fantasy
2,3,Grumpier Old Men (1995),Comedy|Romance
3,4,Waiting to Exhale (1995),Comedy|Drama|Romance
4,5,Father of the Bride Part II (1995),Comedy


<class 'pandas.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   movieId  9742 non-null   int64
 1   title    9742 non-null   str  
 2   genres   9742 non-null   str  
dtypes: int64(1), str(2)
memory usage: 626.1 KB

Basic Info : 
 None
Shape :  (9742, 3)

Columns :  ['movieId', 'title', 'genres']

DataTypes : 
 movieId    int64
title        str
genres       str
dtype: object

Missing Value : 
 movieId    0
title      0
genres     0
dtype: int64

Duplicate Rows :  0

No. of Genres :  951

Genres type :  ['Adventure', 'Animation', 'Children', 'Comedy', 'Fantasy', 'Romance', 'Drama', 'Action', 'Crime', 'Thriller', 'Horror', 'Mystery', 'Sci-Fi', 'War', 'Musical', 'Documentary', 'IMAX', 'Western', 'Film-Noir', '(no genres listed)']


genres
Drama                 4361
Comedy                3756
Thriller              1894
Action                1828
Romance               1596
Adventure             1263
Crime                 1199
Sci-Fi                 980
Horror                 978
Fantasy                779
Children               664
Animation              611
Mystery                573
Documentary            440
War                    382
Musical                334
Western                167
IMAX                   158
Film-Noir               87
(no genres listed)      34
Name: count, dtype: int64

In [52]:
print("=="*50)
print("Ratings Dataset")
print("=="*50)

print("First 5 Rows : ")
display(ratings.head())
print("\nBasic Info : \n",ratings.info())
print("Shape : ",ratings.shape)
print("\nColumns : ",ratings.columns.tolist())
print("\nDataTypes : \n",ratings.dtypes)
print("\nMissing Value : \n",ratings.isnull().sum())
print("\nDuplicate Rows : ",ratings.duplicated().sum())
print("\nNo. of users :",ratings["userId"].nunique())
print("\nNo. of ratings :",sorted(ratings["rating"].unique()))

Ratings Dataset
First 5 Rows : 


,userId,movieId,rating,timestamp
0,1,1,4.0,964982703
1,1,3,4.0,964981247
2,1,6,4.0,964982224
3,1,47,5.0,964983815
4,1,50,5.0,964982931


<class 'pandas.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB

Basic Info : 
 None
Shape :  (100836, 4)

Columns :  ['userId', 'movieId', 'rating', 'timestamp']

DataTypes : 
 userId         int64
movieId        int64
rating       float64
timestamp      int64
dtype: object

Missing Value : 
 userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

Duplicate Rows :  0

No. of users : 610

No. of ratings : [np.float64(0.5), np.float64(1.0), np.float64(1.5), np.float64(2.0), np.float64(2.5), np.float64(3.0), np.float64(3.5), np.float64(4.0), np.float64(4.5), np.float64(5.0)]


In [32]:
ratings.describe()

,userId,movieId,rating,timestamp
count,100836.000000,100836.000000,100836.000000,1.008360e+05
mean,326.127564,19435.295718,3.501557,1.205946e+09
std,182.618491,35530.987199,1.042529,2.162610e+08
min,1.000000,1.000000,0.500000,8.281246e+08
25%,177.000000,1199.000000,3.000000,1.019124e+09
50%,325.000000,2991.000000,3.500000,1.186087e+09
75%,477.000000,8122.000000,4.000000,1.435994e+09
max,610.000000,193609.000000,5.000000,1.537799e+09


In [43]:
print("=="*50)
print("Links Dataset")
print("=="*50)

print("First 5 Rows : ")
display(links.head())
print("\nBasic Info : \n",links.info())
print("Shape : ",links.shape)
print("\nColumns : ",links.columns.tolist())
print("\nDataTypes : \n",links.dtypes)
print("\nMissing Value : \n",links.isnull().sum())
print("\nDuplicate Rows : ",links.duplicated().sum())

Links Dataset
First 5 Rows : 


,movieId,imdbId,tmdbId
0,1,114709,862.0
1,2,113497,8844.0
2,3,113228,15602.0
3,4,114885,31357.0
4,5,113041,11862.0


<class 'pandas.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   movieId  9742 non-null   int64  
 1   imdbId   9742 non-null   int64  
 2   tmdbId   9734 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 228.5 KB

Basic Info : 
 None
Shape :  (9742, 3)

Columns :  ['movieId', 'imdbId', 'tmdbId']

DataTypes : 
 movieId      int64
imdbId       int64
tmdbId     float64
dtype: object

Missing Value : 
 movieId    0
imdbId     0
tmdbId     8
dtype: int64

Duplicate Rows :  0


In [34]:
links.describe()

,movieId,imdbId,tmdbId
count,9742.000000,9.742000e+03,9734.000000
mean,42200.353623,6.771839e+05,55162.123793
std,52160.494854,1.107228e+06,93653.481487
min,1.000000,4.170000e+02,2.000000
25%,3248.250000,9.518075e+04,9665.500000
50%,7300.000000,1.672605e+05,16529.000000
75%,76232.000000,8.055685e+05,44205.750000
max,193609.000000,8.391976e+06,525662.000000


In [51]:
print("=="*50)
print("Tags Dataset")
print("=="*50)

print("First 5 Rows : ")
display(tags.head())
print("\nBasic Info : \n",tags.info())
print("Shape : ",tags.shape)
print("\nColumns : ",tags.columns.tolist())
print("\nDataTypes : \n",tags.dtypes)
print("\nMissing Value : \n",tags.isnull().sum())
print("\nDuplicate Rows : ",tags.duplicated().sum())
print("\nDifferent tags : ",tags["tag"].nunique())
print("\nUsers count : ",tags["userId"].nunique())

Tags Dataset
First 5 Rows : 


,userId,movieId,tag,timestamp
0,2,60756,funny,1445714994
1,2,60756,Highly quotable,1445714996
2,2,60756,will ferrell,1445714992
3,2,89774,Boxing story,1445715207
4,2,89774,MMA,1445715200


<class 'pandas.DataFrame'>
RangeIndex: 3683 entries, 0 to 3682
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   userId     3683 non-null   int64
 1   movieId    3683 non-null   int64
 2   tag        3683 non-null   str  
 3   timestamp  3683 non-null   int64
dtypes: int64(3), str(1)
memory usage: 151.7 KB

Basic Info : 
 None
Shape :  (3683, 4)

Columns :  ['userId', 'movieId', 'tag', 'timestamp']

DataTypes : 
 userId       int64
movieId      int64
tag            str
timestamp    int64
dtype: object

Missing Value : 
 userId       0
movieId      0
tag          0
timestamp    0
dtype: int64

Duplicate Rows :  0

Different tags :  1589

Users count :  58


In [36]:
tags.describe()

,userId,movieId,timestamp
count,3683.000000,3683.000000,3.683000e+03
mean,431.149335,27252.013576,1.320032e+09
std,158.472553,43490.558803,1.721025e+08
min,2.000000,1.000000,1.137179e+09
25%,424.000000,1262.500000,1.137521e+09
50%,474.000000,4454.000000,1.269833e+09
75%,477.000000,39263.000000,1.498457e+09
max,610.000000,193565.000000,1.537099e+09


## Observations

### Movies Dataset

- Number of movies: 9742
- Number of genres: 20
- Genres type :  ['Adventure', 'Animation', 'Children', 'Comedy', 'Fantasy', 'Romance', 'Drama', 'Action', 'Crime', 'Thriller', 'Horror', 'Mystery', 'Sci-Fi', 'War', 'Musical', 'Documentary', 'IMAX', 'Western', 'Film-Noir', '(no genres listed)']
- Missing values: 0
- Duplicate rows: 0

### Ratings Dataset

- Number of ratings: 100836
- Number of users who did rating: 610
- Rating scale: [0.5,1.0,1.5,2.0,2.5,3.0,3.5,4.0,4.5,5.0]
- Missing values: 0

### Links Dataset

- Purpose of this dataset: Links Dataset is used to map MovieLens and TMDB datasets 
- Missing values: 8 - tmdbId (i.e 8 tmdb links)

### Tags Dataset

- Number of tags: 1589
- Users count :  58
- Missing values: 0 
- Interesting observations: Tags are user-generated textual descriptions that capture semantic information about movies. They can boost content-based recommendations by providing additional context beyond genres.